README.md
# Convert some data into Panda DataFrames

## British Petroleum (BP)

It parse sheet like `Primary Energy Consumption` (not like `Primary Energy - Cons by fuel`).

Open: http://www.bp.com/statisticalreview
or https://www.bp.com/en/global/corporate/energy-economics/statistical-review-of-world-energy.html

Download `Statistical Review of World Energy – all data`.

Use:

```python
from shifter_pandas.bp import UNITS_ENERGY, BPDatasource

shifter_ds = BPDatasource("bp-stats-review-2021-all-data.xlsx")

df = shifter_ds.datasource(units_filter=UNITS_ENERGY, regions_filter=["Switzerland"])
df
```

## Swiss Office Federal of Statistics (OFS)

From https://www.bfs.admin.ch/bfs/fr/home/services/recherche/stat-tab-donnees-interactives.html
create a stat table.

Click on `À propos du tableau`

Click on `Rendez ce tableau disponible dans votre application`

Use:

```python
from shifter_pandas.ofs import OFSDatasource

shifter_ds = OFSDatasource("<URL>")

df = shifter_ds.datasource(<Requête Json>)
df
```

And replace `<URL>` and `<Requête Json>` with the content of the fields of the OFS web page.

### Interesting sources

- [Parc de motocycles par caractéristiques techniques et émissions](https://www.pxweb.bfs.admin.ch/pxweb/fr/px-x-1103020100_165/-/px-x-1103020100_165.px/)
- [Bilan démographique selon l'âge et le canton](https://www.pxweb.bfs.admin.ch/pxweb/fr/px-x-0102020000_104/-/px-x-0102020000_104.px/)

## Our World in Data

Select a publication.

Click `Download`.

Click `Full data (CSV)`.

Use:

```python
import pandas as pd
from shifter_pandas.wikidata_ import WikidataDatasource

df_owid = pd.read_csv("<file name>")
wdds = WikidataDatasource()
df_wd = wdds.datasource_code(wikidata_id=True, wikidata_name=True, wikidata_type=True)
df = pd.merge(df_owid, df_wd, how="inner", left_on='iso_code', right_on='Code')
df
```

### Interesting sources

- [GDP, 1820 to 2018](https://ourworldindata.org/grapher/gdp-world-regions-stacked-area)
- [Population, 1800 to 2021](https://ourworldindata.org/grapher/population-since-1800)

## World Bank

Open https://data.worldbank.org/

Find a chart

In `Download` click `CSV`

Use:

```python
from shifter_pandas.worldbank import wbDatasource

df = wbDatasource("<file name>")
df
```

### Interesting sources

- [GDP (current US$)](https://data.worldbank.org/indicator/NY.GDP.MKTP.CD)
- [GDP (constant 2015 US$)](https://data.worldbank.org/indicator/NY.GDP.MKTP.KD)

## Wikidata

By providing the `wikidata_*` parameters, you can ass some data from WikiData.

Careful, the WikiData is relatively slow then the first time you run it il will be slow.
We use a cache to make it fast the next times.

You can also get the country list with population and ISO 2 code with:

```python
from shifter_pandas.wikidata_ import (
    ELEMENT_COUNTRY,
    PROPERTY_ISO_3166_1_ALPHA_2,
    PROPERTY_POPULATION,
    WikidataDatasource,
)

shifter_ds = WikidataDatasource()
df = shifter_ds.datasource(
    instance_of=ELEMENT_COUNTRY,
    with_id=True,
    with_name=True,
    properties=[PROPERTY_ISO_3166_1_ALPHA_2, PROPERTY_POPULATION],
    limit=1000,
)
df
```


CHANGELOG.md
# Changelog

## [0.2.0](https://github.com/sbrunner/shifter-pandas/tree/0.2.0) (2022-03-08)

[Full Changelog](https://github.com/sbrunner/shifter-pandas/compare/0.1.0...0.2.0)

**Implemented enhancements:**

- Add capacity sources, fix unit conversion [\#18](https://github.com/sbrunner/shifter-pandas/pull/18) ([sbrunner](https://github.com/sbrunner))

**Merged pull requests:**

- Bump prospector from 1.7.4 to 1.7.6 [\#17](https://github.com/sbrunner/shifter-pandas/pull/17) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump c2cciutils from 1.1.3 to 1.1.4 [\#16](https://github.com/sbrunner/shifter-pandas/pull/16) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump types-requests from 2.27.10 to 2.27.11 [\#15](https://github.com/sbrunner/shifter-pandas/pull/15) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump prospector from 1.6.0 to 1.7.4 [\#14](https://github.com/sbrunner/shifter-pandas/pull/14) ([dependabot[bot]](https://github.com/apps/dependabot))

## [0.1.0](https://github.com/sbrunner/shifter-pandas/tree/0.1.0) (2022-02-26)

[Full Changelog](https://github.com/sbrunner/shifter-pandas/compare/894a900206d1b42ed2d850b112ab8f3360703eda...0.1.0)

**Implemented enhancements:**

- Add units conversion from BP [\#11](https://github.com/sbrunner/shifter-pandas/pull/11) ([sbrunner](https://github.com/sbrunner))
- Add Worldbank [\#9](https://github.com/sbrunner/shifter-pandas/pull/9) ([sbrunner](https://github.com/sbrunner))
- Add wikidata [\#8](https://github.com/sbrunner/shifter-pandas/pull/8) ([sbrunner](https://github.com/sbrunner))
- Add Our World in Data in readme [\#5](https://github.com/sbrunner/shifter-pandas/pull/5) ([sbrunner](https://github.com/sbrunner))
- Add BP and OFS [\#3](https://github.com/sbrunner/shifter-pandas/pull/3) ([sbrunner](https://github.com/sbrunner))

**Fixed bugs:**

- Little fix [\#12](https://github.com/sbrunner/shifter-pandas/pull/12) ([sbrunner](https://github.com/sbrunner))
- Fix region matching [\#10](https://github.com/sbrunner/shifter-pandas/pull/10) ([sbrunner](https://github.com/sbrunner))
- Fix [\#2](https://github.com/sbrunner/shifter-pandas/pull/2) ([sbrunner](https://github.com/sbrunner))

**Merged pull requests:**

- Use constants [\#13](https://github.com/sbrunner/shifter-pandas/pull/13) ([sbrunner](https://github.com/sbrunner))
- Bump c2cciutils from 1.1.0 to 1.1.3 [\#7](https://github.com/sbrunner/shifter-pandas/pull/7) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump pylint from 2.11.1 to 2.12.2 [\#6](https://github.com/sbrunner/shifter-pandas/pull/6) ([dependabot[bot]](https://github.com/apps/dependabot))
- Make it locally installable [\#4](https://github.com/sbrunner/shifter-pandas/pull/4) ([sbrunner](https://github.com/sbrunner))

\* _This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)_


tests/test_bp.py
"""
Tests of BP Datasource.
"""


from shifter_pandas.bp import UNITS_ENERGY, BPDatasource


def test_bp() -> None:
    """
    Tests of BP Datasource.
    """

    shifter_ds = BPDatasource("tests/bp-stats-review-2021-all-data.xlsx")
    metadata = shifter_ds.metadata()
    assert metadata[1]["regions"][:5] == [
        {"index": 5, "label": "Canada"},
        {"index": 6, "label": "Mexico"},
        {"index": 7, "label": "US"},
        {"index": 8, "label": "Total North America"},
        {"index": 10, "label": "Argentina"},
    ]
    del metadata[1]["regions"]
    assert metadata[1]["years"][:5] == [
        {"index": 2, "label": 1965},
        {"index": 3, "label": 1966},
        {"index": 4, "label": 1967},
        {"index": 5, "label": 1968},
        {"index": 6, "label": 1969},
    ]
    del metadata[1]["years"]

    assert metadata[3]["regions"][:2] == [
        {"index": 6, "label": "Mexico"},
        {"index": 7, "label": "US"},
    ]
    del metadata[3]["regions"]
    assert metadata[3]["years"][:2] == [
        {"index": 2, "label": 1975},
        {"index": 3, "label": 1980},
    ]
    del metadata[3]["years"]

    assert metadata == [
        {"index": 0, "supported": False, "type": "Contents"},
        {
            "index": 1,
            "label": "Primary Energy Consumption",
            "row_index": 3,
            "supported": True,
            "type": "Primary Energy Consumption",
            "unit": {
                "iso": "J",
                "iso_factor": 1000000000000000000,
                "iso_postfix": "",
                "normalized": "exajoules",
                "original": "Exajoules",
            },
        },
        {"index": 2, "supported": False, "type": "Cobalt and Lithium - Prices"},
        {
            "index": 3,
            "label": "Geothermal Capacity",
            "row_index": 4,
            "supported": True,
            "type": "Geothermal Capacity",
            "unit": {
                "iso": "W",
                "iso_factor": 1000000,
                "iso_postfix": "",
                "normalized": "megawatts",
                "original": "Megawatts",
            },
        },
        {"index": 4, "supported": False, "type": "Approximate conversion factors"},
    ]

    assert shifter_ds.to_iso_unit == {
        "lb": {"unit": "tonnes", "factor": 2204.62},
        "short tons": {"unit": "tonnes", "factor": 1.1023},
        "barrels": {"unit": "m³", "factor": 6.2898},
        "litres": {"unit": "m³", "factor": 1000},
        "kilocalorie": {"unit": "J", "factor": 0.0041841004184100415},
        "kcal": {"unit": "J", "factor": 0.0041841004184100415},
        "calorie": {"unit": "J", "factor": 4.184100418410042},
        "cal": {"unit": "J", "factor": 4.184100418410042},
        "Btu": {"unit": "J", "factor": 0.0010548523206751054},
        "barrel of oil equivalent": {"unit": "j", "factor": 6119000},
        "kilowatt-hour": {"unit": "J", "factor": 3.6},
        "kilowatt-hours": {"unit": "J", "factor": 3.6},
        "kWh": {"unit": "J", "factor": 3.6},
        "watt-hour": {"unit": "J", "factor": 3600},
        "watt-hours": {"unit": "J", "factor": 3600},
        "Wh": {"unit": "J", "factor": 3600},
        "cubic meters": {"unit": "m³", "factor": 1},
        "meters": {"unit": "m", "factor": 1},
        "joules": {"unit": "J", "factor": 1},
        "cubic feets": {"unit": "m³", "factor": 35.3146667},
        "cubic meter": {"unit": "m³", "factor": 1},
        "meter": {"unit": "m", "factor": 1},
        "joule": {"unit": "J", "factor": 1},
        "cubic feet": {"unit": "m³", "factor": 35.3146667},
        "watts": {"unit": "W", "factor": 1},
        "watt": {"unit": "W", "factor": 1},
        "us gallons": {"unit": "m³", "factor": 0.14969724},
        "gallons": {"unit": "m³", "factor": 0.14969724},
    }

    data_frame = shifter_ds.datasource(regions_filter=["Total World"], years_factor=20)
    assert set(data_frame.Year) == {1980, 2000, 2020}
    assert set(data_frame.Region) == {"Total World"}
    assert set(data_frame.TypeUnit) == {"Geothermal Capacity [W]", "Primary Energy Consumption [J]"}
    assert set(data_frame.Type) == {"Geothermal Capacity", "Primary Energy Consumption"}
    assert set(data_frame.Unit) == {"W", "J"}

    data_frame = shifter_ds.datasource(regions_filter=["Total World", "Switzerland"], years_factor=50)
    assert set(data_frame.Region) == {"Total World", "Switzerland"}

    data_frame = shifter_ds.datasource(
        regions_filter=["Total World", "Switzerland"], years_filter=[1935, 1945]
    )
    assert set(data_frame.Year) == set()

    data_frame = shifter_ds.datasource(
        regions_filter=["Total World", "Switzerland"], years_filter=[1985, 1993]
    )
    assert set(data_frame.Year) == {1985, 1993}

    data_frame = shifter_ds.datasource(types_filter=["Geothermal Capacity"], years_factor=50)
    assert set(data_frame.Type) == {"Geothermal Capacity"}

    data_frame = shifter_ds.datasource(units_filter=UNITS_ENERGY, years_factor=50)
    assert set(data_frame.Type) == {"Primary Energy Consumption"}


tests/test_ofs.py
# https://www.bfs.admin.ch/bfs/fr/home/statistiques/catalogues-banques-donnees/donnees.assetdetail.18904904.html

from shifter_pandas.ofs import OFSDatasource


def test_ofs():
    shifter_ds = OFSDatasource(
        "https://www.pxweb.bfs.admin.ch/api/v1/fr/px-x-0204000000_106/px-x-0204000000_106.px"
    )

    data_field = shifter_ds.datasource(
        {
            "query": [
                {
                    "code": "Wirtschaft und Haushalte",
                    "selection": {"filter": "item", "values": ["1"]},
                }
            ],
            "response": {"format": "json-stat"},
        }
    )
    assert list(data_field.columns) == [
        "values",
        "Unité de mesure",
        "Economie et ménages",
        "Agent énergétique",
        "Année",
    ]
    data_field.rename(columns={"Année": "Year"}, inplace=True)
    assert set(data_field.Year) == {str(e) for e in range(2000, 2020)}


tests/test_wikidata.py
# https://data.worldbank.org/indicator/NY.GDP.MKTP.KD

from shifter_pandas.wikidata_ import PROPERTY_ISO_3166_1_ALPHA_2
from shifter_pandas.worldbank import WorldbankDatasource


def test_wikidata() -> None:
    shifter_ds = WorldbankDatasource("tests/API_NY.GDP.MKTP.KD_DS2_en_csv_v2_3630701.zip")
    data_field = shifter_ds.datasource(
        wikidata_id=True,
        wikidata_name=True,
        wikidata_type=True,
        wikidata_properties=[PROPERTY_ISO_3166_1_ALPHA_2],
    ).query("CountryName in ['OECD members', 'Switzerland', 'World', 'European Union', 'Euro area']")
    assert list(data_field.columns) == [
        "Year",
        "Value",
        "CountryName",
        "CountryCode",
        "IndicatorName",
        "IndicatorCode",
        "WikidataType",
        "WikidataId",
        "WikidataName",
        "WikidataIso3166_1Alpha_2Code",
    ]
    assert set(data_field.WikidataType) == {None, "country", "World"}
    assert set(data_field.WikidataId) == {None, "Q16502", "Q30", "Q39"}
    assert set(data_field.WikidataName) == {None, "United States of America", "world", "Switzerland"}
    assert set(data_field.WikidataIso3166_1Alpha_2Code) == {None, "US", "CH"}


shifter_pandas/ofs.py
"""Datasource builder for data from the swiss Office Federal of Statistics."""

from typing import Any, Dict, List, Optional, cast

import pandas as pd
import requests

from shifter_pandas.wikidata_ import ELEMENT_CANTON_CH, WikidataDatasource


# https://www.bfs.admin.ch/bfs/fr/home/services/recherche/api/api-pxweb.html
class OFSDatasource:
    """Datasource builder for data from the swiss Office Federal of Statistics."""

    def __init__(self, url: str) -> None:
        """Initialize the datasource builder."""
        self.url = url
        self.wdds = WikidataDatasource()

    def metadata(self) -> Dict[str, Any]:
        """Get the metadata."""
        response = requests.get(self.url, timeout=120)
        if not response.ok:
            print(f"Error on query {self.url}: {response.status_code}")
            print(response.text)
            response.raise_for_status()
        return cast(Dict[str, Any], response.json())

    def datasource(
        self,
        query: Dict[str, Any],
        wikidata_dimension: Optional[str] = None,
        wikidata_id: bool = False,
        wikidata_name: bool = False,
        wikidata_properties: Optional[List[str]] = None,
    ) -> pd.DataFrame:
        """Get the Datasource as DataFrame."""

        if wikidata_properties is None:
            wikidata_properties = []
        wikidata = wikidata_id or wikidata_name or wikidata_properties

        response = requests.post(self.url, json=query, timeout=120)
        if not response.ok:
            print(f"Error on query {self.url}: {response.status_code}")
            print(response.text)
            response.raise_for_status()

        json = response.json()

        values = {"values": json["dataset"]["value"]}
        length = 1
        total_length = len(json["dataset"]["value"])
        for dimension_id in json["dataset"]["dimension"]["id"]:
            dimension = json["dataset"]["dimension"][dimension_id]

            current_length = len(dimension["category"]["index"])
            number = int(total_length / (length * current_length))

            dimension_value = list(json["dataset"]["value"])
            for index_x in range(length):
                for index_y, value in enumerate(dimension["category"]["label"].values()):
                    for index_z in range(number):
                        dimension_value[
                            index_x * current_length * number + index_y * number + index_z
                        ] = value

            values[dimension["label"]] = dimension_value

            length *= current_length

        if wikidata and wikidata_dimension:

            def _get_values(canton: str) -> Dict[str, Any]:
                element_ids = self.wdds.get_from_alias(ELEMENT_CANTON_CH, canton)

                element = {}
                if element_ids:
                    element.update(
                        self.wdds.get_item(
                            element_ids[0]["id"],
                            with_name=wikidata_name,
                            properties=wikidata_properties,
                            with_id=wikidata_id,
                            prefix="Wikidata",
                        )
                    )

                return element

            data = [_get_values(canton) for canton in values[wikidata_dimension]]

            if wikidata_id:
                values["WikidataId"] = [item.get("WikidataId") for item in data]
            if wikidata_name:
                values["WikidataName"] = [item.get("Name") for item in data]
            for wikidata_property in wikidata_properties:
                wikidata_property_name = self.wdds.get_property_name(wikidata_property)
                values[wikidata_property_name] = [item.get(wikidata_property_name) for item in data]

        return pd.DataFrame(values)


shifter_pandas/bp.py
"""Datasource builder for data from British Petroleum."""

from typing import Any, Dict, List, Optional, Tuple

import openpyxl
import pandas as pd

from shifter_pandas import standardize_property
from shifter_pandas.wikidata_ import WikidataDatasource

# ISO
UNITS_ENERGY = ["J", "J (input-equivalent)"]
# ISO
UNITS_ENERGY_PER_CAPITA = ["J per capita"]
# ISO
UNITS_VOLUME = ["m³"]
# ISO
UNITS_VOLUME_PRE_DAY = ["m³ per day"]
# ISO
UNITS_MASS = ["tonnes"]


class BPDatasource:
    """Datasource builder for data from British Petroleum."""

    def __init__(self, file_name: str) -> None:
        """Initialize the datasource builder."""
        self.xlsx = openpyxl.load_workbook(file_name)
        self.wdds = WikidataDatasource()
        self.wdds.set_alias("World", "World", "Q16502", "World")
        # Crude oil: oil_units_conversion[<from unit>][<to unit>] = <factor>
        self.oil_units_conversion: Dict[str, Dict[str, float]] = {}
        # Oil products: oil_products_units_conversion[<product>][<from unit>][<to unit>] = <factor>
        self.oil_products_units_conversion: Dict[str, Dict[str, Dict[str, float]]] = {}
        # Natural gas (NG) and liquefied natural gas (LNG):
        # gaz_units_conversion[<from unit>][<to unit>] = <factor>
        self.gaz_units_conversion: Dict[str, Dict[str, float]] = {}

        # to_iso_unit[<from_unit>] = {"unit": <to_unit>, "factor": <factor>}
        # tones is considered as ISO because it's more convenient to use
        # then J, m³, tonnes (day, capita).
        self.to_iso_unit: Dict[str, Dict[str, Any]] = {}

        # 1 metric tonne = 2204.62 lb.
        self.to_iso_unit["lb"] = {"unit": "tonnes", "factor": 2204.62}
        # = 1.1023 short tons
        self.to_iso_unit["short tons"] = {"unit": "tonnes", "factor": 1.1023}
        # 1 kilolitre = 6.2898 barrels
        self.to_iso_unit["barrels"] = {"unit": "m³", "factor": 6.2898}
        # 1 kilolitre = 1 cubic meter
        self.to_iso_unit["litres"] = {"unit": "m³", "factor": 1000}
        # 1 kilocalorie (kcal) = 4.1868 kJ = 3.968 Btu
        # 1 kilojoule (kJ) = 1,000 joules = 0.239 kcal = 0.948 Btu
        self.to_iso_unit["kilocalorie"] = {"unit": "J", "factor": 1 / 239}
        self.to_iso_unit["kcal"] = self.to_iso_unit["kilocalorie"]
        self.to_iso_unit["calorie"] = {"unit": "J", "factor": 1 / 0.239}
        self.to_iso_unit["cal"] = self.to_iso_unit["calorie"]
        self.to_iso_unit["Btu"] = {"unit": "J", "factor": 1 / 948}
        # 1 British thermal unit (Btu) = 0.252 kcal = 1.055 kJ
        # 1 barrel of oil equivalent (boe) = 5.8 million Btu = 6.119 million kJ
        self.to_iso_unit["barrel of oil equivalent"] = {"unit": "j", "factor": 6119000}
        # 1 kilowatt-hour (kWh) = 860 kcal = 3600 kJ = 3412 Btu
        self.to_iso_unit["kilowatt-hour"] = {"unit": "J", "factor": 3600 / 1000}
        self.to_iso_unit["kilowatt-hours"] = self.to_iso_unit["kilowatt-hour"]
        self.to_iso_unit["kWh"] = self.to_iso_unit["kilowatt-hour"]
        self.to_iso_unit["watt-hour"] = {"unit": "J", "factor": 3600}
        self.to_iso_unit["watt-hours"] = self.to_iso_unit["watt-hour"]
        self.to_iso_unit["Wh"] = self.to_iso_unit["watt-hour"]

        self.to_iso_unit["cubic meters"] = {"unit": "m³", "factor": 1}
        self.to_iso_unit["meters"] = {"unit": "m", "factor": 1}
        self.to_iso_unit["joules"] = {"unit": "J", "factor": 1}
        self.to_iso_unit["cubic feets"] = {"unit": "m³", "factor": 35.3146667}
        self.to_iso_unit["cubic meter"] = self.to_iso_unit["cubic meters"]
        self.to_iso_unit["meter"] = self.to_iso_unit["meters"]
        self.to_iso_unit["joule"] = self.to_iso_unit["joules"]
        self.to_iso_unit["cubic feet"] = self.to_iso_unit["cubic feets"]
        self.to_iso_unit["watts"] = {"unit": "W", "factor": 1}
        self.to_iso_unit["watt"] = self.to_iso_unit["watts"]

        for type_index, type_value in enumerate(self.xlsx.sheetnames):
            if type_value == "Approximate conversion factors":
                self.units_sheet = self.xlsx.worksheets[type_index]

        for raw in range(8, 13):
            from_unit = self.normalize_unit(self.units_sheet.cell(raw, 1).value)
            from_iso_unit, _, from_iso_factor = self._iso_unit(from_unit)
            for col in range(4, 9):
                to_unit_1 = self.units_sheet.cell(4, col).value
                to_unit = self.normalize_unit(
                    ((to_unit_1 + " ") if to_unit_1 is not None else "") + self.units_sheet.cell(5, col).value
                )
                to_iso_unit, _, to_iso_factor = self._iso_unit(to_unit)
                value = self.units_sheet.cell(raw, col).value
                if isinstance(value, (int, float)) and from_unit != to_unit:
                    self.oil_units_conversion.setdefault(from_unit, {})[to_unit] = value
                    self.oil_units_conversion.setdefault(from_iso_unit, {})[to_iso_unit] = (
                        value / from_iso_factor * to_iso_factor
                    )

        self.to_iso_unit["us gallons"] = {
            "unit": "m³",
            "factor": self.oil_units_conversion["us gallons"]["m³"],
        }
        self.to_iso_unit["gallons"] = self.to_iso_unit["us gallons"]

        for raw in range(20, 27):
            product = self.units_sheet.cell(raw, 1).value
            for col in range(3, 9):
                from_unit = self.normalize_unit(self.units_sheet.cell(16, col).value)
                from_iso_unit, _, from_iso_factor = self._iso_unit(from_unit)
                to_unit = self.normalize_unit(self.units_sheet.cell(17, col).value[2:])
                to_iso_unit, _, to_iso_factor = self._iso_unit(to_unit)
                value = self.units_sheet.cell(raw, col).value
                if isinstance(value, (int, float)) and from_unit != to_unit:
                    self.oil_products_units_conversion.setdefault(product, {}).setdefault(from_unit, {})[
                        to_unit
                    ] = value
                    self.oil_products_units_conversion.setdefault(product, {}).setdefault(from_iso_unit, {})[
                        to_iso_unit
                    ] = (value / from_iso_factor * to_iso_factor)

        for raw in range(33, 39):
            from_unit = self.normalize_unit(self.units_sheet.cell(raw, 1).value)
            from_iso_unit, _, from_iso_factor = self._iso_unit(from_unit)
            for col in range(3, 10):
                to_unit = self.normalize_unit(
                    self.units_sheet.cell(29, col).value + " " + self.units_sheet.cell(30, col).value
                )
                to_iso_unit, _, to_iso_factor = self._iso_unit(to_unit)
                value = self.units_sheet.cell(raw, col).value
                if isinstance(value, (int, float)) and from_unit != to_unit:
                    self.gaz_units_conversion.setdefault(from_unit, {})[to_unit] = value
                    self.gaz_units_conversion.setdefault(from_iso_unit, {})[to_iso_unit] = (
                        value / from_iso_factor * to_iso_factor
                    )

    @staticmethod
    def normalize_unit(unit: str) -> str:
        """Get normalized unit."""
        unit = unit.strip()
        unit = unit.lower()
        if unit.startswith("1 "):
            unit = unit[2:]
        unit = unit.replace("/", " / ")
        unit = unit.replace("  ", " ")
        unit = unit.replace("equiv.", "equivalent")
        unit = unit.replace(" daily", " per day")
        return unit

    def _iso_unit(self, unit: str) -> Tuple[str, str, float]:
        if "/" not in unit:
            return self._single_iso_unit(unit)
        else:
            upper, lower = unit.split("/")
            unit, upper_postfix, factor = self._single_iso_unit(upper.strip())
            assert not upper_postfix, f"Upper unit {upper} has a postfix {upper_postfix}"
            lower_unit, postfix, lower_factor = self._single_iso_unit(lower.strip())
            return f"{unit} / {lower_unit}", postfix, factor / lower_factor

    def _single_iso_unit(self, unit: str) -> Tuple[str, str, float]:
        unit_postfix = ""
        postfix = ""
        for postfix_candidate in ("*",):
            if unit.endswith(postfix_candidate):
                postfix = postfix_candidate
                unit = unit[: -len(postfix_candidate)]
                break

        for postfix_candidate in (
            " (input-equivalent)",
            " per capita",
            " daily",
            " per year",
            " per day",
        ):
            if unit.endswith(postfix_candidate):
                unit_postfix = postfix_candidate
                unit = unit[: -len(postfix_candidate)]
                break
        for postfix_candidate in (
            " of carbon dioxide",
            " of oil equivalent",
            " of lithium content",
            "1",
            "*",
        ):
            if unit.endswith(postfix_candidate):
                postfix = postfix_candidate
                unit = unit[: -len(postfix_candidate)]
                break

        factor = 1
        if unit.startswith("kilo"):
            unit = unit[4:]
            factor = 1000
        elif unit.startswith("mega"):
            unit = unit[4:]
            factor = 1000000
        elif unit.startswith("giga"):
            unit = unit[4:]
            factor = 1000000000
        elif unit.startswith("tera"):
            unit = unit[4:]
            factor = 1000000000000
        elif unit.startswith("peta"):
            unit = unit[4:]
            factor = 1000000000000000
        elif unit.startswith("exa"):
            unit = unit[3:]
            factor = 1000000000000000000
        elif unit.startswith("thousand million "):
            unit = unit[17:]
            factor = 1000000000
        elif unit.startswith("thousand "):
            unit = unit[9:]
            factor = 1000
        elif unit.startswith("million "):
            unit = unit[8:]
            factor = 1000000
        elif unit.startswith("billion "):
            unit = unit[8:]
            factor = 1000000000
        elif unit.startswith("trillion "):
            unit = unit[9:]
            factor = 1000000000000

        if unit in self.to_iso_unit:
            definition = self.to_iso_unit[unit]
            unit = definition["unit"]
            factor *= definition["factor"]

        return unit + unit_postfix, postfix, factor

    def metadata(self) -> List[Dict[str, Any]]:
        """Get the metadata."""
        metadata: List[Dict[str, Any]] = []
        for type_index, type_value in enumerate(self.xlsx.sheetnames):
            nice_type = type_value
            for postfix in (
                " - TWh",
                " - EJ",
                " - PJ",
                " - Cons capita",
                " - Barrels",
                " - Tonnes",
                " - Kboed",
                " - Prices",
            ):
                if type_value.endswith(postfix):
                    nice_type = type_value[: -len(postfix)]

            row_index = -1
            for index in (3, 4):
                value = self.xlsx.worksheets[type_index].cell(index, 2).value
                if isinstance(value, int) and 1800 < value < 2100:
                    row_index = index
                    break

            if row_index >= 0:
                # Year
                years = []
                index = 2
                while True:
                    value = self.xlsx.worksheets[type_index].cell(row_index, index).value
                    if self.xlsx.worksheets[type_index].cell(row_index - 1, index).value is not None:
                        break
                    years.append({"label": value, "index": index})
                    index += 1

                # Country
                regions = []
                index = row_index + 2
                nb_empty_cells = 0
                while True:
                    value = self.xlsx.worksheets[type_index].cell(index, 1).value
                    if value is not None:
                        nb_empty_cells = 0
                        regions.append({"label": value, "index": index})
                    nb_empty_cells += 1
                    if nb_empty_cells > 5:
                        break
                    index += 1
                unit = {"original": self.xlsx.worksheets[type_index].cell(row_index, 1).value.strip()}
                unit["normalized"] = self.normalize_unit(unit["original"])
                iso_unit, postfix, factor = self._iso_unit(unit["normalized"])
                unit["iso"] = iso_unit
                unit["iso_factor"] = factor
                unit["iso_postfix"] = postfix
                metadata.append(
                    {
                        "type": type_value,
                        "label": nice_type,
                        "index": type_index,
                        "unit": unit,
                        "years": years,
                        "regions": regions,
                        "supported": True,
                        "row_index": row_index,
                    }
                )
            else:
                metadata.append(
                    {
                        "type": type_value,
                        "index": type_index,
                        "supported": False,
                    }
                )

        return metadata

    def datasource(
        self,
        types_filter: Optional[List[str]] = None,
        regions_filter: Optional[List[str]] = None,
        units_filter: Optional[List[str]] = None,
        years_filter: Optional[List[int]] = None,
        years_factor: Optional[int] = None,
        units: str = "iso",
        wikidata_id: bool = False,
        wikidata_type: bool = False,
        wikidata_name: bool = False,
        wikidata_properties: Optional[List[str]] = None,
    ) -> pd.DataFrame:
        """Get the Datasource as DataFrame."""

        if wikidata_properties is None:
            wikidata_properties = []
        wikidata = wikidata_id or wikidata_name or wikidata_properties

        columns = ["Value", "Type", "Unit", "TypeUnit", "Year", "Region"]
        if wikidata:
            if wikidata_id:
                columns.append("WikidataId")
            if wikidata_name:
                columns.append("WikidataName")
            if wikidata_type:
                columns.append("WikidataType")
            for wikidata_property in wikidata_properties:
                columns.append(
                    f"Wikidata{standardize_property(self.wdds.get_property_name(wikidata_property))}"
                )
        data_frame = pd.DataFrame(columns=columns)
        for type_ in self.metadata():
            if not type_["supported"]:
                continue
            type_index = type_["index"]
            type_type = type_["type"]
            type_label = type_["label"]
            years = type_["years"]
            regions = type_["regions"]

            if types_filter is not None and type_type not in types_filter:
                continue

            for year in years:
                if years_filter is not None and year["label"] not in years_filter:
                    continue
                if years_factor is not None and year["label"] % years_factor != 0:
                    continue
                for region in regions:
                    if regions_filter is not None and region["label"] not in regions_filter:
                        continue
                    value = self.xlsx.worksheets[type_index].cell(region["index"], year["index"]).value
                    if not isinstance(value, int) and not isinstance(value, float):
                        continue

                    unit_definition = type_["unit"]
                    unit_postfix = ""
                    if units == "normalized":
                        unit = unit_definition["normalized"]
                    elif units == "iso":
                        unit = unit_definition["iso"]
                        value = value * unit_definition["iso_factor"]
                        unit_postfix = unit_definition["iso_postfix"]
                    else:
                        unit = unit_definition["original"]

                    if units_filter is not None and unit not in units_filter:
                        continue

                    element = {
                        "Value": value,
                        "Year": year["label"],
                        "Region": region["label"],
                        "Type": type_type,
                        "Unit": f"{unit}{unit_postfix}",
                        "TypeUnit": f"{type_label} [{unit}]{unit_postfix}",
                    }
                    if wikidata:
                        region_label = region["label"]
                        if region_label.startswith("Total "):
                            region_label = region_label[6:]
                        element_id = self.wdds.get_region(region_label)
                        element["WikidataType"] = element_id["type"] if element_id else None
                        element.update(
                            self.wdds.get_item(
                                element_id["id"] if element_id else None,
                                with_name=wikidata_name,
                                with_id=wikidata_id,
                                properties=wikidata_properties,
                                prefix="Wikidata",
                            )
                        )
                    data_frame = data_frame.append(element, ignore_index=True)

        return data_frame

    def datasource_non_fossil_electricity_to_primary_energy_factor(
        self, from_year: int = 1900
    ) -> pd.DataFrame:
        """Get the Datasource used to convert non fossil electricity to primary energy as DataFrame."""

        data: Dict[str, List[float]] = {
            "Year": [],
            "Factor": [],
        }
        before_2001 = self.units_sheet["B45"].value
        for year in range(from_year, 2001):
            data["Year"].append(year)
            data["Factor"].append(before_2001)
        for index, year in enumerate(range(2001, 2010)):
            data["Year"].append(year)
            data["Factor"].append(self.units_sheet.cell(index + 46, 2).value)
        for index, year in enumerate(range(2010, 2020)):
            data["Year"].append(year)
            data["Factor"].append(self.units_sheet.cell(index + 45, 5).value)

        return pd.DataFrame(data)
